src/core/SurfaceNode.cc

```diff
@@ -282,33 +284,33 @@ const Geometry *SurfaceNode::createGeometry() const
     double v3 = data[ (i - 1) + (lines - 1) * columns ];
     double v4 = data[ (i) + (lines - 1) * columns ];
 
-    p->append_poly(4);
-    p->insert_vertex(ox + i - 1, oy + 0, min_val);
-    p->insert_vertex(ox + i - 1, oy + 0, v1);
-    p->insert_vertex(ox + i, oy + 0, v2);
-    p->insert_vertex(ox + i, oy + 0, min_val);
-
-    p->append_poly(4);
-    p->append_vertex(ox + i - 1, oy + lines - 1, min_val);
-    p->append_vertex(ox + i - 1, oy + lines - 1, v3);
-    p->append_vertex(ox + i, oy + lines - 1, v4);
-    p->append_vertex(ox + i, oy + lines - 1, min_val);
+    ind1=builder.vertexIndex(Vector3d(ox + i, oy + 0, min_val));
+    ind2=builder.vertexIndex(Vector3d(ox + i, oy + 0, v2));
+    ind3=builder.vertexIndex(Vector3d(ox + i - 1, oy + 0, v1));
+    ind4=builder.vertexIndex(Vector3d(ox + i - 1, oy + 0, min_val));
+    builder.append_poly({ind1,ind2,ind3,ind4});
+
+    ind1=builder.vertexIndex(Vector3d(ox + i - 1, oy + lines - 1, min_val));
+    ind2=builder.vertexIndex(Vector3d(ox + i - 1, oy + lines - 1, v3));
+    ind3=builder.vertexIndex(Vector3d(ox + i, oy + lines - 1, v4));
+    ind4=builder.vertexIndex(Vector3d(ox + i, oy + lines - 1, min_val));
+    builder.append_poly({ind1,ind2,ind3,ind4});
   }
 
   // the bottom of the shape (one less than the real minimum value), making it a solid volume
   if (columns > 1 && lines > 1) {
-    p->append_poly(2 * (columns - 1) + 2 * (lines - 1) );
+    builder.append_poly(2 * (columns - 1) + 2 * (lines - 1) );
     for (int i = 0; i < columns - 1; ++i)
-      p->insert_vertex(ox + i, oy + 0, min_val);
+      builder.prepend_vertex(builder.vertexIndex(Vector3d(ox + i, oy + 0, min_val)));
```

Comment:
1. Maybe use `j` for iterator.
2. Should we share the vertex id when the coordinates are the same, e.g. `i = j`?